File: go/vm.go
package main

import (
	"time"
)

const PerHour = 5

type VM struct {
	StartTime time.Time
	EndTime   time.Time
}

type Spot struct {
	VM
}

type Coster interface {
	Cost() float64
}

func (v VM) Cost() float64 {
	end := v.EndTime

	if end.Equal(time.Time{}) {
		end = time.Now()
	}

	d := float64(end.Sub(v.StartTime)) / float64(time.Hour)
	return d * PerHour
}

func (s Spot) Cost() float64 {
	p := s.VM.Cost()
	return p * 0.8
}

func TotalCost(vms []Coster) float64 {
	total := 0.0
	for _, vm := range vms {
		total += vm.Cost()
	}
	return total
}

File: python/vm.py
from dataclasses import dataclass
from datetime import datetime, timezone

PER_HOUR = 5


@dataclass
class VM:
    start_time: datetime
    end_time: datetime

    def cost(self):
        end_time = self.end_time
        if end_time is None:
            end_time = timezone.utc()

        duration = end_time - self.start_time
        hours = duration.total_seconds() / (60*60)

        return hours * PER_HOUR


class Spot(VM):
    def cost(self):
        p = super().cost()

        return p * .8


def total_cost(vms):
    return sum(vm.cost() for vm in vms)


if __name__ == '__main__':
    vms = [
        VM(start_time=datetime(2022, 4, 12, 17, 30, tzinfo=timezone.utc),
           end_time=datetime(2022, 4, 12, 19, 54, tzinfo=timezone.utc)),
        Spot(start_time=datetime(2022, 4, 13, 9, 46, tzinfo=timezone.utc),
             end_time=datetime(2022, 4, 15, 12, 7, tzinfo=timezone.utc))
    ]

    print(total_cost(vms))
